### ISSUE TITLE
State externalization

### ISSUE THREAD

**Issue by jhellerstein** [hydro-project/hydroflow#929]
Let's get all the state of a transducer exposed for management including checkpoint/restart.
- Relates to #927 inasmuch as lattice "state" is formed via `merge`/cumulatives.
- Is `persist` the same as `merge`?
- Keyed fold followed by join on the same key should *share* state. "Hash teams" in SQL Server. #969 
- What about state that persists across executions (e.g. in storage or a replication library)?
- Relates to the "tables" in Bloom being absent from Hydroflow: is state first-class in the IR and separate from ops?
   - Might be good to have an explicit data model spec

**Comment by MingweiSamuel** [comment]
## handoffs

Handoffs (in the current setup) require whatever you put into them to be `'static`, so no references or any lifetime shenanigans

## state-passing

I.e. in the picture above, the two persists each need to send their state to the join.
I guess just `Rc` does work for that, but that does basically bypasses the state API and any Rust ownership/optimizations

**Comment by MingweiSamuel** [comment]
Maybe a way to do it without using iterators, by making codegen fancier

**Comment by MingweiSamuel** [comment]
## Limitations/problems with current hydroflow implementation

* sharing state between operators
* way for downstream operators to request data from upstream operators
  * or to determine which input has had new input and which input is unchanged
  * sharing state may be a way to solve this (unchanged input could be shared state)
    * is it orthogonal?
      * no argument: unchanged input shouldn't really be getting sent around anyway
      * yes: replaying into datastructures is always bad
      * redundant cached state?
    * or could flag the sent data
* more stuff...

**Comment by MingweiSamuel** [comment]
`persist()` (`lattice_fold()`?) should have two "outputs" - one for deltas and one for the whole state
the one for the whole state is "virtual" - it doesn't need to actually get scheduled - just to share the state reference

**Comment by MingweiSamuel** [comment]
- Should `filter(|_| false)` cause more scheduling?
- Stratum 0 stateful operator causes spinning #961
- Shadaj's filter example

**Comment by MingweiSamuel** [comment]
Symbolic representation of operators when you do `persist() -> filter() -> ...` - if we can't symbolically represent all the operators does does it go back to replay? D:

**Comment by MingweiSamuel** [comment]
Example: State teed to two joins, but on one side we re-key the data - this (seems to) force us to re-materialize one side of the tee

**Comment by MingweiSamuel** [comment]
manual fold:
```rust
use hydroflow_macro::hydroflow_syntax;

#[test]
fn test_manual_fold() {
    let (send, recv) = hydroflow::util::unbounded_channel::<u32>();
    let mut hf = hydroflow_syntax! {
        source_iter([0]) -> my_fold;
        my_fold = union() -> reduce::<'tick>(|a: &mut _, b| *a += b);
        my_fold -> inspect(|x| println!("{}", x)) -> defer_tick_lazy() -> my_fold;

        input -> my_fold;
        input = source_stream(recv);
    };

    println!("L{}", line!());
    send.send(1).unwrap();
    hf.run_tick();
    println!("L{}", line!());
    send.send(2).unwrap();
    send.send(3).unwrap();
    send.send(4).unwrap();
    hf.run_tick();
    println!("L{}", line!());
    send.send(5).unwrap();
    hf.run_tick();
    println!("L{}", line!());
    hf.run_tick();
    println!("L{}", line!());

    hf.run_available();
}
```

**Comment by MingweiSamuel** [comment]
State sharing between persists and child join
![image](https://github.com/hydro-project/hydroflow/assets/6778341/8b9cfb05-4ee7-4f28-8751-79c34753a261)

**Comment by MingweiSamuel** [comment]
Pretty limited what you can do between the state cylinder and the join

**Comment by MingweiSamuel** [comment]
Can we get rid of replaying entirely?

**Comment by MingweiSamuel** [comment]
Maybe with cumulatives passed by value, we can still pass by state API ref - could help with scheduling issues? Just let the operator update the state when it feels like and downstream reads it when it feels like? What about chaining?

**Comment by MingweiSamuel** [comment]
Checkpoint/migration handling moved to #1049

**Comment by MingweiSamuel** [comment]
Implemented in code, ergonomics tbd in #797
